### DISCUSSION TITLE
How to manage multiple dependencies for a single package

### DISCUSSION THREAD

**Discussion by sapetnioc** [brainvisa/brainvisa.github.io#150] [Q&A]
We now have a system allowing to build Conda packagages with dependencies. However, AFAIK, a conda package can have only one set of dependencies. But it can be useful to have different dependencies according to the intended usage of the package. Let's take a simple example. Installing current `capsul` package with conda allows to use Capsul but not to run the tests. Tests requires other modules such as `pytest` and `jupyter` that don't need to be installed for all users. What should be our policy for dependencies ? Should we use a minimal set of dependencies for a package and eventually create other packages to add other dependencies (for instance a package `capsul-test`) ? Or should we include all possibly useful dependencies in a single package ?

**Comment by Hboni** [comment]
IMO, we must keep conda packages light, and install minimal set of dependencies. Having a different package to have more dependencies sounds good to me.
Did you find some examples that have the same mechanism?

**Comment by denisri** [comment]
We would like to allow both but the answer is not that simple: it depeneds on use cases.
1. Developers / testers would like to have test dependencies installed
2. Normal users would like to have all useful dependencies installed simply
3. Developers who are working on custom applications based on a part of brainvisa libs would like to have their needed dependencies minimally installed

Case 2 is probably the most common and neeeds to be simple: a user who wants to run brainvisa/morphologist and anatomist will not bother about installing (for instance and among many others) libjpeg, libminc, openslide, the exact correct combinations of versions of ipython/tornado/jupyter, etc etc when he doesn't even know what are all these libs and which may be optionally used. So maybe we could have packages "with plugins", named "<package>-plugins" which add all optional dependencies (soma-io/aims plugins would install all IO formats libraries, for instance).
Or the other way, the "normal", or "usual" package would install all optional dependencies, and would be based on a "<package>-minimal" package or component, which does not install any optional dependency. Maybe I prefer this second way as it is the most common use case.

**Comment by sapetnioc** [comment]
@Hboni , there are, at least, capsul, populse-bd and soma-workflow that uses alternate dependencies.
It seems that creating one package with minimal dependencies and other packages with extended dependencies is a good idea. We still have to decide what are the names for this package (this discussion include what dependencies are in the main package). I am also in favor of having a main package that "just work" for most cases. To my opinion, this is what is often done on Conda Forge. For instance there are no many packages for Qt, when you install it you get all its components.

To be compatible with what is existing in brainvisa-cmake, I modified casa-distro to create several packages according to what is defined in `EXTRA_REQUIRES` in `info.py`. I added an entry `test` for Capsul, this creates the package `capsul-test` allowing to run the tests (they are passing by the way). But a package cannot be defined only with a list of dependencies ; it requires more metadata such as a summary, etc. So we will also have to decide how to change the extra packages definition and to adapt the code to deploy on PyPi.
